design-proposals/persist-vmi-firmware-uuid.md

```diff
@@ -0,0 +1,181 @@
+a single-phase solution with immediate impact and a two-phased approach that gradually transitions to global uniqueness.
+
+## Single-Phase Solution for Persistent, Universally Unique UUIDs
+
+### 1. Upgrade-Specific Persistence of Firmware.UUID
+
+**Description:** Prior to upgrading KubeVirt, persist the `Firmware.UUID` of existing VMs in `vm.spec`. 
+After the upgrade, any VM without a `Firmware.UUID` in its spec will be considered new.
+For these new VMs, the system defaults to using `vm.metadata.uid` as the firmware UUID if `vm.spec.template.spec.Firmware.UUID`is not defined.
+
+**Pros:**
+- The upgrade-specific code can be safely removed after two or three releases.
+- Simple logic post-upgrade, with a straightforward UUID assignment.
+- Limited disturbance to GitOps workflows, affecting only pre-existing VMs with the current method.
+
+**Cons:**
+- Requires additional code (likely in `virt-operator`) to handle the upgrade-specific persistence.
+- Patching the spec of a running VM may trigger a "restart required" condition.
+
+This solution is effective for establishing UUID persistence and uniqueness within a single release,
+allowing existing VMs to retain their UUID while adopting a globally unique identifier (`vm.metadata.uid`) for new VMs.
+
+## Two-Phased Solution for Persistent, Universally Unique UUIDs
+
+To ensure a truly persistent and universally unique firmware UUID for VMs,
+the two-phased approach provides a gradual transition, balancing workload stability with a smooth progression toward global uniqueness.
+
+### Phase 1: Preserve Current Generation Method and Persist UUID
+
+In Phase 1, we aim to maintain workload stability by preserving the UUID generated with the current method.
+This phase focuses on ensuring VMs retain their UUID across restarts without modifying the existing generation approach.
+Below are potential solutions for this phase:
+
+1. **Persist UUID in the VM’s Spec Field**  
+   **Description:** The UUID would be generated on the VM’s first boot and stored in the VM’s spec field.
+
+   **Pros:**
+   - Straightforward implementation.
+   - Avoids adding a new API field, maintaining backward compatibility.
+
+   **Cons:**
+   - This approach "abuses" the spec by storing data that does not represent the user's desired state.
+   - Could lead to cluttered and harder-to-read VM definitions.
+
+2. **Generate UUID via Webhook and Store in the Spec Field**  
+   **Description:** If no UUID is specified by the user, a webhook generates a random UUID and assigns it to the VM’s spec field.
+
+   **Pros:**
+   - Similar to the first option, avoiding a new API field and maintaining backward compatibility.
+
+   **Cons:**
+   - Adds complexity by involving webhooks, which could slow down performance if many VMs are started simultaneously.
+   - May potentially bottleneck `virt-api`.
+
+3. **Add a Firmware UUID Field to VM Status**  
+   **Description:** Store the generated UUID within the VM’s status.
```

Comment:
Status is only live, it must not be used to store information.

aka any informatoin in the status in only relevant and "persisted" on the runtime platform.
Object smust not be created with a populated status (only empty status is what we want: `status: {}`)

Reply:
> Object smust not be created with a populated status (only empty status is what we want: `status: {}`)

A VM object won't be created with a populated status, but a VM that was created, booted, then shut off would have a non-empty status.

Am I missing anything?

Reply:
ping @fabiand :)
just want to understand if I get you correctly.